js/Field.js

```diff
@@ -0,0 +1,396 @@
+import * as polyfills from'./Polyfills';
+polyfills.installMatches(); //cross browser polyfill for 'matches' (does not supported by IE)
+polyfills.installClosest(); //cross browser polyfill for 'closest' (does not supported by IE)
+polyfills.installCustomEvent(); //cross browser polyfill for 'custom events' (does not supported by IE)
+
+export default class Field{
+    constructor(options) {
+        this._el = options.element;
+        this._width = options.width;
+        this._height = options.height;
+        this._numberOfMines = options.numberOfMines;
+        this._flags = 0;
+
+        this._state = Field.PLAYING;
+
+        this._simultaneousClick = {
+            cell: null,
+            leftButtonDown: false,
+            rightButtonDown: false
+        };
+
+        this._renderField();
+        this._field = this._el.querySelector('[data-selector="field"]');
+        this._minesLeft = this._el.querySelector('[data-selector="minesIndicator"]');
+
+        this._placeMines();
+        this._placeNumbers();
+
+        this._openCell = this._openCell.bind(this);
+        this._field.addEventListener('click',this._openCell);
+
+        this._setFlag = this._setFlag.bind(this);
+        this._field.addEventListener('contextmenu',this._setFlag);
+
+        this._onMouseDown = this._onMouseDown.bind(this);
+        this._field.addEventListener('mousedown',this._onMouseDown);
+
+        this._onMouseUp = this._onMouseUp.bind(this);
+        this._field.addEventListener('mouseup',this._onMouseUp);
+    }
+//------------ static constants describing state of a game-------------
+    static get PLAYING() {
+        return 0;
+    }
+    static get END_OF_GAME() {
+        return 1;
+    }
+
+    //-----------get methods---------------
+    get width() {
+        return this._width;
+    }
+    get height() {
+        return this._height;
+    }
+    get numberOfMines() {
+        return this._numberOfMines;
+    }
+    get flags() {
+        return this._flags;
+    }
+    get state() {
+        return this._state;
+    }
+
+//----------------public methods--------------------
+    clear() {
+        this._el.removeChild(this._field);
+        this._field = null;
+    }
+    cellNeighbours(y,x) {
+        let neighbours = [];
+
+        for (let i = Math.max(y-1, 0); i <= Math.min(y+1, this.height-1); i++) {
+            for (let j = Math.max(x-1, 0); j <= Math.min(x+1, this.width-1); j++) {
+                if ( !(i === y && j === x) ) {
+                    neighbours.push(this._field.rows[i].cells[j]);
+                }
+            }
+        }
+
+        return neighbours;
+    }
+
+    //-----------------event listeners----------------
+    _openCell(event) {
+        if (this.state === Field.END_OF_GAME) {
+            return null;
+        }
+
+        let td = event.target.closest('td');
+
+        if (!td) {
+            return null;
+        }
+
+        if (td.dataset.type !== 'covered') {
+            return null;
+        }
```

Comment:
you've mixed event handling with real job
try to avoid it because method became to heavy